File: 0001-0500/0088.Merge Sorted Array.py
class Solution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: None Do not return anything, modify nums1 in-place instead.
        """
        p = m + n - 1
        m -= 1
        n -= 1
        while n != -1:
            if m == -1 or nums2[n] >= nums1[m]:
                nums1[p] = nums2[n]
                n -= 1
            else:
                nums1[p] = nums1[m]
                m -= 1
            p -= 1

File: 0001-0500/0088.Merge Sorted Array.c
/*
Basic idea:
	For oridinary mergesort, we merge from the heading. However, in this problem the heading array is temporarily used.
	Since the array is increasingly sorted from head to tail, it is equivalent to an array decreasingly sorted from tail to head. We can merge from tail to head, and plcae the result from tail to head.

59 / 59 test cases passed.
Status: Accepted
Runtime: 0 ms
Your runtime beats 10.49% of c submissions.
*/
void merge(int* nums1, int m, int* nums2, int n) {
    int target = m+n-1;
    m--;n--;
    while(m>=0 && n>=0)
        if (nums1[m]>nums2[n])
            nums1[target--]=nums1[m--];
        else
            nums1[target--]=nums2[n--];
    if (n>=0)
        memcpy(nums1, nums2, sizeof(int)*(n+1));
}

